README.md
PhysUnits-RT
============

A C++ library for run-time dimensional analysis and unit/quantity manipulation and conversion.

This library is based on the quantity compile-time library by Michael S. Kenniston[1] and adapted and expanded for runtime use by Martin Moene.

This easy to use header-only library is intended to work with a wide range of compilers, including MSVC 6.

Example
---------
```C++
#include "phys/units/quantity.hpp"

using namespace phys::units;

int main()
{
    quantity speed = 45 * kilo() * meter() / second();
}
```

<h2>Other libraries</h2>
- [PhysUnits-CT](https://github.com/martinmoene/PhysUnits-CT) - Compile-time companion of this library.
- [Boost.Units](http://www.boost.org/doc/libs/1_51_0/libs/units/) - Zero-overhead dimensional analysis and unit/quantity manipulation and conversion in C++.

Usage
-------
### Definition of terms

Adapted from Boost.Units:
- <b>Base dimension</b>: A base dimension is loosely defined as a measurable entity of interest; in conventional dimensional analysis, base dimensions include length ([L]), mass ([M]), time ([T]), etc.. Base dimensions are essentially a tag type and provide no dimensional analysis functionality themselves.
- <b>Dimension</b>: A collection of zero or more base dimensions, each potentially raised to a different rational power. For example, length = [L]^1, area = [L]^2, velocity = [L]^1/[T]^1, and energy = [M]^1 [L]^2/[T]^2 are all dimensions.
- <b>Base unit</b>: A base unit represents a specific measure of a dimension. For example, while length is an abstract measure of distance, the meter is a concrete base unit of distance. Conversions are defined using base units. Much like base dimensions, base units are a tag type used solely to define units and do not support dimensional analysis algebra.
- <b>Unit</b>: A set of base units raised to rational exponents, e.g. m^1, kg^1, m^1/s^2.
- <b>System</b>: A unit system is a collection of base units representing all the measurable entities of interest for a specific problem. For example, the SI unit system defines seven base units : length ([L]) in meters, mass ([M]) in kilograms, time ([T]) in seconds, current ([I]) in amperes, temperature ([theta]) in kelvin, amount ([N]) in moles, and luminous intensity ([J]) in candelas. All measurable entities within the SI system can be represented as products of various integer or rational powers of these seven base units.
- <b>Quantity</b>: A quantity represents a concrete amount of a unit. Thus, while the meter is the base unit of length in the SI system, 5.5 meters is a quantity of length in that system.

### Limitations
This library only supports the use of the <em>SI unit system</em>. Note however that the library does provide a fixed number (10) of extension dimensions that you can assign names to.

This library only supports <em>integral powers</em> of the dimensions.

The <em>representation or value type</em> in the implementation of quantity is fixed and cannot be defined separately for each quantity. However you can change the type for all uses within a translation unit by defining `PHYS_UNITS_REP_TYPE` before inclusion of header quantity.hpp. Default this type is `double`.

### Output

The following example shows the quantity type in the computation of work from force and distance and the printing of the result on standard output.
```C++
#include <iostream>

#include "phys/units/io.hpp"
#include "phys/units/quantity.hpp"

using namespace phys::units;
using namespace phys::units::io;

quantity
work( const quantity & F, const quantity & dx )
{
    return F * dx; // Defines the relation: work = force * distance.
}

int main()
{
    quantity      F ( 2.0 * newton());  // Define a quantity of force.
    quantity      dx( 2.0 * meter() );  // and a distance,
    quantity      E ( work( F, dx ) );  // and calculate the work done.

    std::cout << "F  = " << F  << std::endl
              << "dx = " << dx << std::endl
              << "E  = " << E  << std::endl;
}
```

The output produced is:
```
F  = 2 N
dx = 2 m
E  = 4 J
```

The following example demonstrates printing in default floating point notation and in [engineering notation](http://en.wikipedia.org/wiki/Engineering_notation), using [metric prefixes](http://en.wikipedia.org/wiki/Metric_prefix).
```C++
#include <iostream>

#include "phys/units/io.hpp"
#include "phys/units/quantity.hpp"

using namespace phys::units;

int main()
{
    quantity R( 4.7 * kilo() * volt() / ampere() );

    {
        using namespace phys::units::io;
        std::cout << "R = " << R << std::endl;
    }
    {
        using namespace phys::units::io::eng;
        std::cout << "R = " << R << std::endl;
    }
}
```
The output produced is:
```
R = 4700 Ohm
R = 4.7 kOhm
```
See namespaces `io` and `io::eng` for further information.

Instead of unit names such as J, you can also obtain the unit expressed in base dimensions.
```C++
#include <iostream>

#include "phys/units/io.hpp"
#include "phys/units/quantity.hpp"

using namespace phys::units;
using namespace phys::units::io;

int main()
{
    std::cout << "J = " << to_base_unit_symbols( joule() ) << std::endl;
}
```
The output produced is:
```
J = m+2 kg s-2
```

### Unit name extension
There are ten extension dimensions available: `extra_x0_d..extra_x9_d`. You can define names for these dimensions on a first come, first serve basis. The following example demonstrates the extension with a new unit (dimension) named `Foo`.
```C++
#include <iostream>

#include "phys/units/io.hpp"
#include "phys/units/quantity.hpp"

using namespace phys::units;
using namespace phys::units::io;

int main()
{
    quantity O( unit( "Ohm"           ) );  // the default is no_extend()
    quantity F( unit( "Foo", extend() ) );  // extend() enables extension needed here

    std::cout << "O = " << O << std::endl
              << "F = " << F << std::endl;
}
```
The output produced is:
```
O = 1 Ohm
F = 1 Foo
```
If you only want to make a limited number of self-defined unit names available, you can predefine these via for example `unit("Foo", extend())` and omit parameter `extend()` from further calls to `unit()` and `to_quantity()`.

### Input

A textual specification of a quantity can be converted to a quantity via function `to_quantity()` as the following examples demonstrates.

```C++
#include <iostream>

#include "phys/units/io.hpp"
#include "phys/units/quantity.hpp"

using namespace phys::units;
using namespace phys::units::io;

int main()
{
    std::cout << "quantity = " << to_quantity( "3 J/A" )                   << std::endl
              << "quantity = " << to_quantity( "4"     , dimensionless() ) << std::endl
              << "quantity = " << to_quantity( "5 Foo" , extend()        ) << std::endl;
}
```
The output produced is:
```
quantity = 3 Wb
quantity = 4
quantity = 5 Foo
```

Recognized quantity formats are for example: 1 m, 1 m2, 1 m/s, 1 m/s2, 1 (4 cm)2, 33 Hz, 33 1/s. If the magnitude is omitted, it is taken to be 1. To prevent misinterpretation with self-defined unit names, an escape character can be used as in: !foo , f!foo (same as ffoo).

Internally `to_quantity()` uses class `quantity_parser`.

### Convenience functions

There are several convenience functions, such as:
- `to_quantity(std::string text)` - the text, such as '9.8 m/s2' converted to quantity.
- `to_unit(std::string text)`- the text, such as '45 km/h' converted to quantity representing the unit as '1 km/h'.
- `to_numerical_value(std::string text)` - the text, such as '45 km/h' converted to value as '45'
- `magnitude(quantity const & q)` - the quantity's magnitude, e.g. 1000 for 1 km.
- `dimension(quantity const & q)` - the quantity's dimension, e.g. m/s for 1 km/h.
- `to_unit_symbol(quantity const &q, name_symbol use)` - the quantity's unit symol(s), for example 'm s-2'.
- `to_real(quantity const &q)` - a dimensionless quantity's magnitude as floating point value.
- `to_integer(quantity const &q)` - a dimensionless quantity's magnitude as integral value.

### Error handling

Exceptions are used to signal errors in expressions that use incompatible dimensions and for several error conditions related to handling the textual representation of a quantity.

The following example demonstrates simple error handling that prints the message provided by the exception.
```C++
#include <iostream>

#include "phys/units/io.hpp"
#include "phys/units/quantity.hpp"

using namespace phys::units;
using namespace phys::units::io;

int main()
{
    try
    {
        std::cout << "quantity = " << to_quantity( "3 m/s/s" ) << std::endl;
    }
    catch( std::exception const & e )
    {
        std::cout << "Error: " << e.what() << std::endl;
    }
}
```
The output produced is:
```
Error: quantity: parsing unit '3 m/s/s': repetition of solidus at position 6: use parenthesis to avoid ambiguity
```
For an example that uses `quantity_parser_error` from the quantity exception hierarchy, see the error handling in the program [PhysUnits.cpp](https://svn.eld.leidenuniv.nl/repos/eld-moene/PhysUnitsRT/trunk/projects/PhysUnits/PhysUnits.cpp), function `void interactive()`.

### Mixing run-time and compile-time libraries
Although this library seems to be in the `phys::units` namespace, it does so through *using* the `rt` namespace in the global namespace. So this library actually lives in the `rt::phys::units` namespace. When you want to mix this library with its compile-time companion, the `rt` namespace must be applied. This can be accomplished by defining `PHYS_UNITS_IN_RT_NAMESPACE`.

Dependencies
--------------
This library has no dependencies other than the standard C++ libraries.

Performance
-------------
```
Relative running time (lower is better)

Compiler        Option : double : quantity
-----------------------+--------+-------------
GCC 4.6.2         -O2  :  1     :  222 (1)
MS VC6/VS6        -O2  :  0.83  :  396 (1.8)
MS VC8/VS2005     -O2  :  0.83  :  269 (1.2)
MS VC2010/VS2010  -O2  :  0.66  :  391 (1.8)

Measured on a AMD Athlon 64 X2 Dual Core Processor 5600+, 64kB L1 Data, 64kB L1 Instruction, 512kB L2, 3.2 GB RAM
```

Compilers known to work
-------------------------
- GCC 4.5.2, 4.6.2
- MS VC6/VS6 - (VC6 lacks ADL).
- MS VC8/VS2005
- MS VC2010/VS2010

Ideas for improvement
-----------------------
Currently the number of extension dimensions is fixed. Perhaps the code can be made to work so that the number of extension dimensions can be specified via a user-defined value of `PHYS_UNITS_QUANTITY_UNIT_EXT_COUNT`.

Allow to specify a conversion offset between two units, e.g. to make conversion between 'C and K possible (see Boost.Units).

It may be nice if you can obtain a quantity in a unit representation of your choice, e.g. in kWh in stead of J [m+2 kg s-2]. See G.S. Novak. [Conversion of units of measurement (PDF)](http://reference.kfupm.edu.sa/content/c/o/conversion_of_units_of_measurement__42431.pdf)". 1 August 1997.

### References
[1] Michael Kenniston. [The Quantity Library](http://home.xnet.com/~msk/quantity/quantity.html). ([Rationale](http://home.xnet.com/%7Emsk/quantity/quantity.html), Quantity [folder](http://www.xnet.com/%7Emsk/quantity)). 16 July 2001, rev 0.4.

[2] Ambler Thompson and Barry N. Taylor. [Guide for the Use of the International System of Units (SI)](http://physics.nist.gov/cuu/pdf/sp811.pdf). NIST Special Publication 811 2008 Edition.


ReleaseNotes.txt
Release Notes for Runtime Quantity C++ Library.


Release 1.0.0 - 2012-10-18

First public release.

Features:
- Added converting a text representation of a quantity or unit to a quantity.
- Added formatting of non-base units, such as Ohm, m/s.
- Added extending dimensions with user-defined dimensions/units.
- Added doxygen documentation.
- Changed tests to use Catch test framework by Phil Nash,
  https://github.com/philsquared/Catch (VS2008, VS2010, GCC),
  https://github.com/martinmoene/Catch (VS6 and above)


Rev. 0.4 - 2001-07-16

Original library by Michael S. Kenniston.

___

Features
Bug Fix


projects/PhysUnits/PhysUnits.cpp
/**
 * \file PhysUnits.cpp
 *
 * \brief   Console program to play with units.
 * \author  Martin Moene
 * \date    16 September 2012
 *
 * Copyright 2012 Universiteit Leiden. All rights reserved.
 *
 * Distributed under the Boost Software License, Version 1.0. (See accompanying
 * file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
 */

// g++ -Wall -O2 -I../ -I%BOOST_INCLUDE% -o PhysUnits.exe PhysUnits.cpp && PhysUnits "1.23 m" "330 m/s" "9.8 m/s2"
// cl -nologo -W3 -EHsc -D_CRT_SECURE_NO_WARNINGS -I../ -I%BOOST_INCLUDE% PhysUnits.cpp && PhysUnits "1.23 m" "330 m/s" "9.8 m/s2"

#define PHYS_UNITS_MICRO_GLYPH "u"

#include "phys/units/io.hpp"
#include "phys/units/quantity.hpp"

#include <iostream>
#include <cstring>

#ifdef PHYS_UNITS_COMPILER_IS_MSVC
# include <io.h>
#else
# include <stdio.h>
# include <unistd.h>
#endif

char const * const version_text =
    "1.4 (12 October 2012)";

char const * const copyright_text =
    "Copyright 2012, Universiteit Leiden";

char const * const freesoftware_text =
    "This is free software, based on work by Michael S. Kenniston\n"
    "and distributed under the Boost Software License version 1.0.";

inline const std::string to_string( const long x, const int base = 10 )
{
   char buf[65];
   return _ltoa( x, buf, base );
}

/**
 * return compiler name, version.
 */
std::string compiler_name()
{
#if defined( _MSC_VER )
    return "MSVC " + to_string(_MSC_VER / 100 - 6) + "." + to_string(_MSC_VER % 100) ;
#elif defined( __GNUC__ )
    return "GCC " + to_string(__GNUC__) + "." + to_string(__GNUC_MINOR__) + "." + to_string(__GNUC_PATCHLEVEL__) ;
#endif
    return "Unknown";
}

/**
 * print version text, return EXIT_SUCCESS.
 */
int version( std::string const progname )
{
    std::cout <<
        progname << " " << version_text << " " << compiler_name() << "\n" <<
        copyright_text  << "\n" <<
        freesoftware_text <<  std::endl;
    return EXIT_SUCCESS;
}

/**
 * print help text, return EXIT_SUCCESS.
 */
int usage( std::string const progname )
{
    std::cout <<
        "\nUsage: " << progname << " [option...] [--] [-] [\"expression\"...]" <<
        "\n"
        "\nParse the unit expressions given on the commandline or read from standard"
        "\ninput and present each resulting quantity in engineering notation and as"
        "\nunit expressed in the seven SI base units."
        "\n"
        "\nOptions"
        "\n      -h, --help: this help."
        "\n   -d, --dimless: accept dimensionless quantities."
        "\n  -e, --escape=c: define escape character for extended units [!]."
        "\n    -x, --extend: define units when first encountered [no extend]."
        "\n         --debug: report debug info [no debug]."
        "\n       --version: report version of program."
        "\n              --: end option section."
        "\n               -: read from standard input."
        "\n"
        "\nOption --escape implies option --extend."
        "\n"
        "\nExamples"
        "\n  " << progname << " \"42 km\" \"1 dm3\" \"2 (3.14 mm)2\""
        "\n  " << progname << " \"330 m/s\" \"9.8 m/s2\" \"9.8 m.s-2\" \"9.8 m s-2\""
        "\n  " << progname << " \"3 kHz\" \"3 1/s\" \"3 kV.A\" \"2.2 kOhm\""
        "\n  " << progname << " --extend \"3 Foo\" \"4 !foo\" \"ffoo\" \"f!foo\" \"J2/ffoo\""
        "\n"
        "\nSyntax (EBNF)"
        "\n     expression = [magnitude] factor { (\" \"|\".\"|\"/\") factor } ."
        "\n"
        "\n         factor = prefixed-unit [power]"
        "\n                | \"(\" expression \")\" ."
        "\n"
        "\n  prefixed-unit = [prefix] unit"
        "\n"
        "\n      magnitude = floating-point-number"
        "\n"
        "\n          power = signed-integral-number"
        "\n"
        "\n         prefix = \"y\"..\"Y\""
        "\n"
        "\n           unit = [\"!\"] (\"m\"|\"kg\"|\"s\"|\"A\"|\"K\"|\"mol\"|\"cd\"...)"
        "\n                |  \"1\""
        "\n"
        "\nNote 1: \"!\" is the default escape character for newly defined units."
        "\nNote 2: \"1\" is used for reciprocal units; it must be followed by \"/\"."
        "\n"
        "\nFor more information on SI units, see:"
        "\n  Ambler Thompson and Barry N. Taylor."
        "\n  Guide for the Use of the International System of Units (SI)."
        "\n  NIST Special Publication 811 2008 Edition."
        "\n  http://physics.nist.gov/cuu/pdf/sp811.pdf"

    << std::endl;

    return EXIT_SUCCESS;
}

/**
 * print error messag and return EXIT_FAILURE.
 */
int error( std::string msg )
{
    std::cerr << "Error: " << msg << std::endl;
    return EXIT_FAILURE;
}

/**
 * check for and return single character from option argument.
 */
int expect_character( char const * const option_str, char const * const value_str )
{
    if ( strlen( value_str ) != 1 )
    {
        exit( error( std::string() +
            "expecting single character for option '" + option_str + "', got '" + value_str + "'\n"
            "Try option --help for more information." ) );
    }
    return value_str[0];
}

/**
 * report text and quantity created from that.
 */
void report( std::string const argument, phys::units::quantity const q )
{
    using namespace phys::units;
    using phys::units::io::eng::operator<<;

    std::cout << "'" << argument  << "': " << q <<
        " [" << to_unit_symbol( q, base_symbols() ) << "]" << std::endl;
}

/**
 * true if stream represents a console.
 * \sa http://stackoverflow.com/questions/5156675/how-to-use-isatty-on-cout-or-can-i-assume-that-cout-file-descriptor-1
 */
bool isConsole( std::istream & is )
{
    return 0 != _isatty( _fileno( stdin ) );
}

/**
 * read a line of text.
 * VC6 doesn't handle end of line correctly with std::cin, hence a separate version.
 */
std::string readline()
{
#ifdef PHYS_UNITS_COMPILER_IS_MSVC6
    std::string text;

    do
    {
        if ( isConsole( std::cin ) )
        {
            std::cout << ">";
        }

        char buf[200];
        char * const p = fgets( buf, sizeof buf - 1, stdin );

        if ( NULL == p || feof( stdin ) )
        {
            return "exit";
        }

        text = p;
        text.erase( text.size() - 1 );
    }
    while ( text.size() <= 0 );

    return text;
#else
    std::cin.clear();
    std::string text;

    while( text.size() <= 0 )
    {
        if ( isConsole( std::cin ) )
        {
            std::cout << ">";
        }

        std::getline( std::cin, text );

        if ( std::cin.eof() )
        {
            return "exit";
        }
    }
    return text;
#endif
}

/**
 * interactive mode.
 */
void interactive( std::string const progname, phys::units::quantity_parser & parser )
{
    if ( isConsole( std::cin ) )
    {
        std::cout <<
            "\n" << progname << " " << version_text << " " << compiler_name() <<
            "\n"
            "\nCommands: help, exit (or Ctrl+Z)\n" << std::endl;
    }

    while ( true )
    {
        try
        {
            std::string expression = readline();

            if ( expression == "help" )
            {
                (void) usage( progname );
            }
            else if ( expression == "exit" )
            {
                return ;
            }
            else
            {
                report( expression, parser.parse( expression ) );
            }
        }
        catch( phys::units::quantity_parser_error const & e)
        {
            std::cout <<
               e.input_text() <<
               "\n" << std::string( e.input_position() - 1, ' ' ) << "^" <<
               "\nError: " << e.what() << std::endl;
        }
    }
}

/**
 * main program.
 */
int main( int argc, char *argv[] )
{
    try
    {
        char const * const progname = "PhysUnits" ; // argv[0];

        bool opt_debug = false;
        bool opt_extend = false;
        bool opt_dimless = false;
        char opt_escape = '!';

        // process options:

        while( *++argv && '-' == *argv[0] )
        {
            // argument such as: -a=1.23, --alpha=1.23, --trace
            char * const argument = argv[0];

            /*
             * separate option from (optional) value:
             *
             * Note value may always be accessed, but can yield an empty string ("")
             * if nothing was specified (e.g. as in --anoption ).
             */
            char *pos = strchr(argument, '=');
            if ( NULL != pos )
            {
                *pos++ = '\0';
            }
            else
            {
                // Point to end of string / empty string;
                // saves us allocating a seperate one.
                pos = (char *) "";
            }

            char const * const option = argument;   // e.g. -a or --alpha
            char const * const value  = pos;        // e.g. 1.23, or ""

            if ( strcmp( "--", option) == 0 )
            {
                ++argv;
                break;
            }
            else if ( strcmp( "-h", option) == 0 || strcmp( "--help", option ) == 0 )
            {
                return usage( progname );
            }
            else if ( strcmp( "-d", option) == 0 || strcmp( "--dimless", option ) == 0
                                                 || strcmp( "--dimensionless", option ) == 0 )
            {
                opt_dimless = true;
            }
            else if ( strcmp( "-e", option) == 0 || strcmp( "--escape", option ) == 0 )
            {
                opt_escape = expect_character( option, value );
                opt_extend = true;
            }
            else if ( strcmp( "-x", option) == 0 || strcmp( "--extend", option ) == 0 )
            {
                opt_extend = true;
            }
            else if ( strcmp( "--debug", option ) == 0 )
            {
                opt_debug = true;
            }
            else if ( strcmp( "--version", option ) == 0 )
            {
                return version( progname );
            }
            else
            {
                return error( std::string() +
                    "invalid option '" + option + "'.\n"
                    "Try option --help for more information." );
            }
        }

        // no arguments, show help:

        if ( 0 == *argv )
        {
            return usage( progname );
        }

        // process arguments:

        while ( *argv )
        {
            using namespace phys::units;

            std::string const argument( *argv++ );

            quantity_parser parser;

            parser.
                debug ( opt_debug  ).
                extend( opt_extend ).
                escape( opt_escape ).
                dimensionless( opt_dimless );

            if ( "-" == argument )
            {
                interactive( progname, parser );
            }
            else
            {
                report( argument, parser.parse( argument ) );
            }
        }
    }
    catch( std::exception const & e)
    {
        std::cout << "Error: " << e.what() << std::endl;
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS; // VC6
}

/*
 * end of file
 */
